### ISSUE TITLE
Getting a "cannot pickle" error in Wagtail 5.1 when when using a custom image model

### ISSUE THREAD

**Issue by DanielSwain** [wagtail/wagtail#10879]
### Issue Summary

I recently upgraded four small sites to Wagtail 5.1.  None of these sites have ever used caching, and each site uses a custom image model.  On all but one site I started getting the error: ` cannot pickle '_io.BufferedReader' object` every place an `{% image ... %}` tag is used.  All of these sites have been working fine for several years.  (As an aside, I see that in `Image.get_rendition()`, a `try...except` was removed in 5.1 that would have probably been masking any caching-related errors that might have occurred before 5.1.)  The pickling error occurs both when caching is absent and when caching (Redis) is used.  The `'_io.BufferedReader' object` error occurs in Django's `locmem.py` caching when no cache is specified in the settings, and a different pickling error message occurs when Redis is used, but in both cases pickling is the problem.

To test from the command line, I inserted a print statement in `locmem.py` to print out the pickled value.  I then looped through the images calling `get_rendition()`.  All pickled values printed correctly.  To test a specific image from the command line, I uploaded a new image to the `logo_admin` field described below and did:
```
r=CustomRendition.objects.get(pk=63)
r.image.get_rendition(r.image_spec)
```
The pickled value was printed correctly.  However, when this exact same image/rendition was called from an `{% image ... %}` tag in the admin as described below, the pickling error happens.  The `CustomImage(AbstractImage)/CustomRendition(AbstractRendition)` definitions are the same in each of these projects and are set up [as described in the docs](https://docs.wagtail.org/en/latest/advanced_topics/images/custom_image_model.html#custom-image-models).

The easiest way to test is from the admin as described above and in **Steps to Reproduce** below.  I also tried:
- deleting all existing renditions
- uploading a new image to the `logo_admin` field shown below.  Again, when I comment out the `{% image settings.main.SiteSettings.logo_admin width-150 %}` line shown below, the admin comes up fine.  When I uncomment it, the error happens again.  The error happens with both `png` and `jpg` images.

### Main Question
How could the exact same rendition be pickled fine when called from the admin but encounter a pickling error when called from an `{% image ... %}` tag??

### Steps to Reproduce

Use a `CustomImage` model and set up the following `SiteSettings` model:

```
@register_setting
class SiteSettings(BaseSiteSetting):
    logo_admin = models.OneToOneField(settings.WAGTAILIMAGES_IMAGE_MODEL, null=True, blank=True, on_delete=models.SET_NULL, related_name='+')
```

Render the following from the `base.html` template in `wagtailadmin`:

```
{% extends "wagtailadmin/base.html" %}
{% load wagtailimages_tags %}

{% block branding_logo %}
    {% image settings.main.SiteSettings.logo_admin width-150 %}
{% endblock %}
```

-   I have NOT confirmed that this issue can be reproduced as described on a fresh Wagtail project.

### Technical details

-   Python version:  3.8
-   Django version:  4.2.5
-   Wagtail version:  5.1

**Comment by zerolab** [comment]
I can confirm this is an issue using Bakerydemo with

```
# local.py

CACHES = {
    "default": {"BACKEND": "django.core.cache.backends.locmem.LocMemCache"},
}

# base/models.py
class SiteSettings(BaseSiteSetting):
    title_suffix = models.CharField(
        verbose_name="Title suffix",
        max_length=255,
        help_text="The suffix for the title meta tag e.g. ' | The Wagtail Bakery'",
        default="The Wagtail Bakery",
    )

    logo_admin = models.OneToOneField(get_image_model(), null=True, blank=True, on_delete=models.SET_NULL, related_name='+')

    panels = [
        FieldPanel("title_suffix"),
        FieldPanel("logo_admin"),
    ]
```

```django=html
{# wagtail/admin/templates/wagtailadmin/base.html #}

{% load wagtailimages_tags %}

{% block branding_logo %}
    {% image settings.base.SiteSettings.logo_admin width-150 %}
{% endblock %}
```

**Comment by gasman** [comment]
Investigating. It seems it's not necessary to use a custom image model, or override the admin base template - it can be replicated by adding the line `{% image settings.base.SiteSettings.logo_admin width-150 %}` to bakerydemo/templates/base/home_page.html and viewing the homepage.

Curiously, preceding that tag with `{% image page.image width-150 %}` - with the homepage's hero image set to the same image as logo_admin in SiteSettings - successfully creates the rendition and allows further calls to `{% image settings.base.SiteSettings.logo_admin width-150 %}` to succeed, until the server is restarted - so apparently there must be something peculiar to the image instance retrieved through SiteSettings that isn't the case for images related to the page.

**Comment by gasman** [comment]
If `logo_admin` is changed from a `OneToOneField` to a `ForeignKey`, the issue does not occur. On the other hand, I'm not able to trigger the issue by changing `HomePage.image` to a `OneToOneField`...

**Comment by gasman** [comment]
If I move the SiteSettings object lookup to Python code:

```
class HomePage(Page):
    # ...
    def get_context(self, request, *args, **kwargs):
        context = super().get_context(request, *args, **kwargs)
        context['logo_admin'] = SiteSettings.for_request(request).logo_admin
        return context
```

with `{% image logo_admin width-150 %}` in the template, the issue does still occur - but not if I change the lookup to `SiteSettings.objects.first().logo_admin`.

**Comment by gasman** [comment]
The problem goes away if this line is removed from `BaseSiteSetting`:

https://github.com/wagtail/wagtail/blob/e6e29b55181a635f6adf7771bd1a7eb808f352a0/wagtail/contrib/settings/models.py#L129-L130

I think I see what's going on now:

* When the `SiteSettings` instance is retrieved through `for_request`, the request object is attached to it as the attribute `_request`. Request objects can't be pickled, so this makes the SiteSettings instance un-pickleable
* That shouldn't be a problem, because we want to pickle a rendition object, not a SiteSettings instance...
* However, the Rendition object has a ForeignKey reference back to the Image, and because the Image was retrieved through a OneToOneField, it has a reverse relation pointing back to the SiteSettings instance. (This isn't the case for a ForeignKey, because the reverse relation is one-to-many, i.e. there isn't a single instance to point back to.)
* Apparently, pickling a Django model instance will cause both of those relations to be followed, and the related objects to be included in the pickled representation. This seems surprising, and might be a Django bug.

In summary:

* The easy workaround here is to change the OneToOneField to a ForeignKey
* We should fix BaseSiteSetting to leave out the `_request` attribute when being pickled, since people might have a legitimate use for pickling those (even if we're only doing it accidentally here)
* We should also investigate whether it's intended behaviour for relations to be followed when pickling a Django model, and report it as a Django bug if not.

**Comment by gasman** [comment]
After a lot of digging around, I found https://code.djangoproject.com/ticket/22330 which seems to match the issue we're seeing here. On that ticket the core team were on the fence about whether holding on to the cached related objects is desirable, and it's hard to see what can be added to the original report that might sway them.

The concerns raised there about performance regressions are valid - the Python [`copy` module](https://docs.python.org/3/library/copy.html) follows the same API for determining what should and shouldn't be preserved, so it's quite likely that plenty of existing Django apps are relying on being able to call `copy(some_model_instance)` and follow cached relations on the object. Given that there's been no further demand for change in the 8 years since that issue was raised, I think it's unlikely that they'll risk changing the status quo.
